crypto/bp/bp_g2.c

```diff
@@ -0,0 +1,1212 @@
+    size_t field_len, i, skip;
+
+    if (form != POINT_CONVERSION_UNCOMPRESSED)
+        goto err;
+
+    if (G2_ELEM_is_at_infinity(group, point)) {
+        /*
+         * encodes to a single 0 octet
+         */
+        if (buf != NULL) {
+            if (len < 1)
+                return 0;
+            buf[0] = 0;
+        }
+        return 1;
+    }
+
+    /*
+     * ret := required output buffer length
+     */
+    field_len = BN_num_bytes(group->field);
+    ret = 4 * field_len;
+
+    /*
+     * if 'buf' is NULL, just return required length
+     */
+    if (buf != NULL) {
+        if (len < ret)
+            goto err;
+
+        if (ctx == NULL && (ctx = new_ctx = BN_CTX_new()) == NULL)
+            return 0;
+
+        used_ctx = 1;
+        BN_CTX_start(ctx);
+        if ((x[0] = BN_CTX_get(ctx)) == NULL
+            || (x[1] = BN_CTX_get(ctx)) == NULL
+            || (y[0] = BN_CTX_get(ctx)) == NULL
+            || (y[1] = BN_CTX_get(ctx)) == NULL)
+            goto err;
+
+        if (!G2_ELEM_get_affine_coordinates(group, point, x, y, ctx))
+            goto err;
+
+        i = 0;
+        skip = field_len - BN_num_bytes(x[0]);
+        if (skip > field_len)
+            goto err;
+        while (skip > 0) {
+            buf[i++] = 0;
+            skip--;
+        }
+        skip = BN_bn2bin(x[0], buf + i);
+        i += skip;
+        if (i != field_len)
+            goto err;
+
+        skip = field_len - BN_num_bytes(x[1]);
+        if (skip > field_len)
+            goto err;
+        while (skip > 0) {
+            buf[i++] = 0;
+            skip--;
+        }
+        skip = BN_bn2bin(x[1], buf + i);
+        i += skip;
+        if (i != 2 * field_len)
+            goto err;
+
+        if (form == POINT_CONVERSION_UNCOMPRESSED) {
+            skip = field_len - BN_num_bytes(y[0]);
+            if (skip > field_len)
+                goto err;
+            while (skip > 0) {
+                buf[i++] = 0;
+                skip--;
+            }
+            skip = BN_bn2bin(y[0], buf + i);
+            i += skip;
+
+            skip = field_len - BN_num_bytes(y[1]);
+            if (skip > field_len)
+                goto err;
+            while (skip > 0) {
+                buf[i++] = 0;
+                skip--;
+            }
+            skip = BN_bn2bin(y[1], buf + i);
+            i += skip;
+        }
+
+        if (i != ret)
+            goto err;
+    }
+
+    if (used_ctx)
+        BN_CTX_end(ctx);
+    BN_CTX_free(new_ctx);
+    return ret;
+
+ err:
+    if (used_ctx)
+        BN_CTX_end(ctx);
+    BN_CTX_free(new_ctx);
+    return 0;
+}
+
+int G2_ELEM_oct2point(const BP_GROUP *group, G2_ELEM *point,
+                      const unsigned char *buf, size_t len, BN_CTX *ctx)
+{
+    BN_CTX *new_ctx = NULL;
+    BIGNUM *x[2], *y[2];
+    size_t field_len, enc_len;
+    int ret = 0;
+
+    if (len == 0)
+        return 0;
+
+    if (len == 1)
+        return G2_ELEM_set_to_infinity(group, point);
```

Comment:
Yes, although it is easy to implement point compression similarly to RFC 5480, since the supported curves have p = 3 mod 4 and thus square roots are simple.